### PR TITLE
Add option to complete slider without animation immediately or from xml attribute

### PR DESCRIPTION
**Pull request by nataliialapshyna-dna**
## Description
This change should fix the issue https://github.com/cortinico/slidetoact/issues/164
Add an extra possibility to set the "complete" view state as the base one from XML.

## Related PRs/Issues
Fixes https://github.com/cortinico/slidetoact/issues/164
Fixes https://github.com/cortinico/slidetoact/issues/52

## Todos
- [x] Tests
- [x] Documentation
- [x] Screenshots

### CODE REVIEW

File: slidetoact/src/main/res/values/attrs.xml [hunk 1]
```diff
@@ -25,6 +25,7 @@
         <attr name="animate_completion" format="boolean" />
         <attr name="text_appearance" format="reference" />
         <attr name="complete_icon" format="reference" />
+        <attr name="complete" format="boolean" />
```
Comment (cortinico) [inline slidetoact/src/main/res/values/attrs.xml:28]: Let's rename this to `state_complete`.

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 2]
```diff
@@ -735,6 +743,31 @@ class SlideToActView @JvmOverloads constructor(
...
+            }
+        }
+        val tickAnimator: ValueAnimator = createIconAnimator(this, mDrawableTick, tickListener)
+        tickAnimator.start()
+    }
+
+    /**
+     * Method for complete slider immediately without animation
+     */
+    fun setCompleted() {
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt:767]:
```suggestion
    fun setCompleted(state: Boolean) {
```

You should also handle state transition back and forth.

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 3]
```diff
@@ -458,6 +460,11 @@ class SlideToActView @JvmOverloads constructor(
 
         // Make sure the position is recomputed.
         mPosition = 0
+
+        // Set state to complete if needed
+        if (mIsCompleted) {
+            setCompleteState()
+        }
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt:467]: I believe this won't work if you call it inside `onSizeChanged`. This should be moved inside the `setCompleted` function

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 4]
```diff
@@ -396,6 +396,8 @@ class SlideToActView @JvmOverloads constructor(
 
                 mArrowMargin = mIconMargin
                 mTickMargin = mIconMargin
+
+                mIsCompleted = getBoolean(R.styleable.SlideToActView_complete, false)
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt:400]: You should check at the end of the constructor what's the value of `mIsCompleted` and call the correct function

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 5]
```diff
@@ -722,6 +729,7 @@ class SlideToActView @JvmOverloads constructor(
 
                 override fun onAnimationEnd(p0: Animator?) {
                     mIsCompleted = true
+                    isEnabled = false
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt:732]: Why this?

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 6]
```diff
@@ -735,6 +743,31 @@ class SlideToActView @JvmOverloads constructor(
...
+        isEnabled = false
+        val tickListener = ValueAnimator.AnimatorUpdateListener {
+            // We need to enable the drawing of the AnimatedVectorDrawable before starting it.
+            if (!mFlagDrawTick) {
+                mFlagDrawTick = true
+                mTickMargin = mIconMargin
+            }
+        }
+        val tickAnimator: ValueAnimator = createIconAnimator(this, mDrawableTick, tickListener)
+        tickAnimator.start()
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt:761]: This should not be animated. The whole point of the `setCompleted` is to reach the completed state without animation essentially :)

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 7]
```diff
@@ -735,6 +743,31 @@ class SlideToActView @JvmOverloads constructor(
...
+
+    /**
+     * Method for complete slider immediately without animation
+     */
+    fun setCompleted() {
+        this.mIsCompleted = true
+    }
+
     /**
      * Method that completes the slider
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt:772]: We do have a API problem here as we have both:

```
fun setCompleted(true)
fun completeSlider()
fun setCompleted(false)
fun resetSlider()
```

Ideally we have something like:

```
fun setCompleted(completed: Boolean, withAnimation: Boolean)

@Deprecate fun completeSlider()
@Deprecate fun resetSlider()
```

**Review by cortinico — COMMENTED** [review]
Thanks for the RP @nataliialapshyna-dna 🙏 
There are several thigns to address but is a step in the right direction

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt (hunk 2 above)
Comment (nataliialapshyna-dna) [reply slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: Now is private. Public fun - setCompleted(completed: Boolean, withAnimation: Boolean)

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt (hunk 4 above)
Comment (nataliialapshyna-dna) [reply slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: Unfortunately, there is a problem.  `setCompleted(mIsCompleted) ` can't be called in constructor as some values like `mAreaWidth`, `mAreaHeight` will be set up only in `onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int)` function.

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt (hunk 6 above)
Comment (nataliialapshyna-dna) [reply slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: Replaced by `startIconAnimation(icon: Drawable)`. I was able to find only such a way to show `mDrawableTick`, but maybe you could advise, how to do it completely without animation.

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt (hunk 7 above)
Comment (nataliialapshyna-dna) [reply slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: `fun setCompleted(completed: Boolean, withAnimation: Boolean) `was added

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt (hunk 5 above)
Comment (nataliialapshyna-dna) [reply slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt:731]: Just leftover, sorry, removed

**Comment by cortinico** [comment]
Sorry for being quite absent @nataliialapshyna-dna. I'll try to re-review this and the other PRs in the near future :)

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt (hunk 4 above)
Comment (cortinico) [reply slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: That also works 👍

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt (hunk 6 above)
Comment (cortinico) [reply slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: I think it's ok to call `startIconAnimation` in this specific section. Agree that is not fully ideal as you would want no animations at all, but this still an improvement from the status quo 👍

**Review by cortinico — APPROVED** [review]
Thank you very much for sending this @nataliialapshyna-dna 🙏 

Sorry it took longer than expected to review, my bad.

**Pull request merged**